reactor-core/src/test/java/reactor/core/publisher/FluxReplayTest.java

```diff
@@ -512,4 +476,56 @@ public void scanSubscriber() {
         test.cancelled = true;
         Assertions.assertThat(test.scan(Scannable.Attr.CANCELLED)).isTrue();
     }
+
+	@Test
+	public void cacheSingleSubscriberWithMultipleRequestsDoesntHang() {
+		await().atMost(5, TimeUnit.SECONDS)
+		       .untilAsserted(() -> assertThat(
+				       Flux.range(0, 1000)
+				           .cache(5)
+				           .toStream(10)
+				           .collect(Collectors.toList()))
+				       .hasSize(1000)
+		       );
+	}
+
+	@Test
+	public void cacheNotOverrunByMaxPrefetch() {
+		Flux<Integer> s = Flux.range(1, 30)
+		                      .cache(5);
+
+		StepVerifier.create(s, 10)
+		    .expectNextCount(10)
+		    .thenRequest(20)
+		    .expectNextCount(20)
+		    .verifyComplete();
+
+		StepVerifier.create(s)
+		            .expectNextCount(5)
+		            .verifyComplete();
+	}
+
+	private static final class TwoRequestsSubscriber extends BaseSubscriber<Integer> {
```

Comment:
Could also be changed to something like `SequentialRequestsSubscriber` or something + be a bit more generic and accept `long... requests`

Reply:
for now it is very ad-hoc for the pattern of tests we have in this class, but it could definitely be changed later on if a more complex pattern emerges. this is test-private for now anyway